### DISCUSSION TITLE
Change Tycho to always use the maven dependecy model

### DISCUSSION THREAD

**Discussion by laeubi** [eclipse-tycho/tycho#818] [Ideas]
With the recent changes in 2.7.x, and the move to maven 3.8.4 we now have an opportunity I'd like to discuss here that seems to me like a very interesting idea in regards of making Tycho more "maven-centric" but will possible be a complete change in the concept of dependency resolution.

Currently it roughly works like this:

1. Tycho resolves for each project the "Target platform", this could use target-files, repositories or (if enabled) declared pom dependencies (aka pomDependecies=consider)
2. The wen inject this (p2) dependencies into the maven model to allow maven to compute a proper build order
3. All Tycho components now work on the "IDependencyMetadata", especially to compute compile/test classpath, build features / products and so on. This currently is a bit a mixture of different items and we do not really distinguish between scopes (e.g. what is test and what is compile scope)

For (2) we already use that information and has enhanced for the `dependency:list` support but the information will not be used any further in tycho itself.

What I like to propose is, that we drop the "IDependencyMetadata" (or better redefine it) and work completely on the maven dependency model calculated at step (2), of course we still will need a mapping of mvn-dependency ->Installable unit but that's only relevant in certain cases.

As an example the compiler plugin can then just scan all dependencies of the pom and choose those that are bundles (or wrap any others if enabled) very similar to how the maven-compiler plugin would behave, and even handle the scopes appropriately (given we inject the correctly).  This of course would mean that `pomDependecies=consider` is active by default, but would actually match what most maven-users would expect and it could simplify some of our tasks as one don't need to care about later stages if a dependency is correctly injected in the first place. There is one (minor) drawback, that maven would require to download all artifacts at the resolve stage but I think if that really hurts I think we should just improve maven to support a lazy model here.

This is obviously something that won't happen today or tomorrow and probably will be painful in some areas as we need to rethink some concepts, but in the end I think will greatly improve how tycho works.

**Comment by mickaelistria** [comment]
What I'm afraid of with `pomDependencies=consider` by default is that some people start having successful build for faulty content (eg Import-Package missing in MANIFEST). But I think that's what compiler access rule are supposed to prevent. Do you confirm that such a case would still fail the build?

**Reply by laeubi** [reply]
Yes this is what the compiler access rules will be used for. And actually there will be cases where access rules are irrelevant because the code will never be used at runtime (e.g. annotations with retention policy of class or source) or the classloaders are merged (fragments) or they are simply not osgi bundles, they should still constitute to the compile classpath (see also https://github.com/eclipse/tycho/issues/793)

**Comment by khatchad** [comment]
Do we still need to follow [these steps](https://wiki.eclipse.org/Tycho/How_Tos/Dependency_on_pom-first_artifacts) to use maven dependencies in tycho projects?
